File: 그래프/이분그래프/baekjoon-1707.py
import sys

sys.setrecursionlimit(10 ** 5)


def dfs(vertex, color, visited, adj_list):
    visited[vertex] = color
    for next in adj_list[vertex]:
        if visited[next] == 0:
            if not dfs(next, 3 - color, visited, adj_list):
                return False
        elif visited[next] == visited[vertex]:
            return False
    return True


if __name__ == "__main__":
    test_case = int(sys.stdin.readline())
    while test_case:
        V, E = map(int, sys.stdin.readline().split())
        adj_list = [[] for _ in range(V + 1)]
        visited = [0] * (V + 1)

        for _ in range(E):
            u, v = map(int, sys.stdin.readline().split())
            adj_list[u].append(v)
            adj_list[v].append(u)

        ans = True
        for start in range(1, V + 1):
            if visited[start] == 0:
                if not dfs(start, 1, visited, adj_list):
                    ans = False
        print("YES" if ans else "NO")

        test_case -= 1

File: 그래프/이분그래프/baekjoon-1707.cpp
#include <cstdio>
#include <cstring>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int color[20001];
vector<int> adj[20001];

bool bfs(int start)
{
  queue<int> q;
  color[start] = 1;
  q.push(start);

  while (!q.empty())
  {
    int now = q.front();
    q.pop();

    for (int next : adj[now])
    {
      if (color[next] == 0)
      {
        q.push(next);
        color[next] = 3 - color[now];
      }
      else if (color[next] == color[now])
      {
        return false;
      }
    }
  }

  return true;
}

int main()
{
  int testCase;
  scanf("%d", &testCase);

  for (size_t t = 0; t < testCase; t++)
  {
    int v, e;
    scanf("%d%d", &v, &e);

    for (size_t i = 1; i <= v; i++)
    {
      adj[i].clear();
      color[i] = 0;
    }

    for (size_t i = 0; i < e; i++)
    {
      int from, to;
      scanf("%d%d", &from, &to);

      adj[from].push_back(to);
      adj[to].push_back(from);
    }

    bool ans = true;
    for (size_t start = 1; start <= v; start++)
    {
      if (color[start] == 0)
      {
        if (!bfs(start))
        {
          ans = false;
        }
      }
    }

    printf("%s\n", ans ? "YES" : "NO");
  }

  return 0;
}